aten/src/ATen/MetaInit.cpp

```diff
@@ -6,23 +6,305 @@
+  TensorImpl* ptr = tensor.unsafeGetTensorImpl();
+
+  auto pos = surrogates_.find(ptr);
+
+  // If we have seen `tensor` as an argument to a previous operation before,
+  // return its cached surrogate since we should preserve its identity.
+  if (pos != surrogates_.end()) {
+    return pos->second;
+  }
+
+  weak_refs_.emplace(ptr);
+
+  // Note that we do not maintain the view relationship of foreign tensors. This
+  // means if two foreign tensors share the same storage, we do not reflect this
+  // in their surrogates. Since we only deal with meta tensors this is fine.
+  Tensor surrogate = empty_like(tensor, device(kMeta));
+
+  surrogates_.emplace(ptr, surrogate);
+
+  return surrogate;
+}
+
+using TensorReplacer = std::function<Tensor(const Tensor&)>;
+
+// Replaces the tensors contained in `value` by calling `fn`.
+IValue replaceTensors(const IValue& value, const TensorReplacer& fn) {
+  bool replaced = false;
+
+  IValue::HashAliasedIValueMap memo{};
+
+  auto visitor = [&replaced, &memo, &fn](const IValue& v) {
+    // Deep-copy all compound objects.
+    if (v.isTuple() || v.isList() || v.isGenericDict()) {
+      return false;
+    }
+
+    if (v.isTensor()) {
+      const Tensor& inp = v.toTensor();
+
+      Tensor out = fn(inp);
+      // If not replaced, we need to put the original value into `memo` so that
+      // it does not get cloned by `deepcopy()`.
+      if (out.is_same(inp)) {
+        memo[v] = v;
+      } else {
+        memo[v] = out;
+
+        replaced = true;
+      }
+    } else {
+      // Do not clone other non-compound objects.
+      memo[v] = v;
+    }
+
+    return false;
+  };
+
+  value.visit(visitor);
+
+  if (replaced) {
+    return value.deepcopy(memo);
+  } else {
+    return value;
+  }
+}
+
+void replaceTensorArguments(const OperatorHandle& op, Stack& s, const TensorReplacer& fn) {
+  std::size_t num_arguments = op.schema().arguments().size();
+
+  // Perform an in-place update of the tensor arguments in the stack.
+  for (std::size_t i = 0; i < num_arguments; i++) {
+    IValue& arg = torch::jit::peek(s, i, num_arguments);
+
+    arg = replaceTensors(arg, fn);
+  }
+}
+
+// Replaces all foreign tensor arguments with surrogate tensors.
+void replaceForeignTensorArguments(const OperatorHandle& op, Stack& s) {
+  auto replace_foreign_tensor = [](const Tensor& tensor) -> Tensor {
+    // If the tensor was not output by an operation that we recorded, we should
+    // replace it with a surrogate tensor.
+    if (tensor.is_meta()) {
+      return tensor;
+    } else {
+      return meta_init_state.getOrCreateSurrogateTensor(tensor);
+    }
+  };
+
+  replaceTensorArguments(op, s, replace_foreign_tensor);
+}
+
+// Checks whether a `TensorOptions` argument can be parsed from the individual
+// arguments of `schema`.
+bool hasTensorOptions(const FunctionSchema& schema) noexcept {
+  std::array<c10::string_view, 4> tensor_opts = {{
+      "dtype", "layout", "device", "pin_memory"
```

Comment:
Is the heuristic here that any one of these should be present in args or all of them?

Reply:
Please see my comment for the line 237. The heuristic is that these parameters should appear in this exact order in the parameter list of the operator to be considered a tensor factory.